File: packages/sdk/java/packages/android/src/main/java/dev/edgebase/sdk/client/FunctionsClient.java
package dev.edgebase.sdk.client;

import dev.edgebase.sdk.core.EdgeBaseError;
import dev.edgebase.sdk.core.HttpClient;

import java.util.Collections;
import java.util.Map;

public class FunctionsClient {
    public static class FunctionCallOptions {
        public final String method;
        public final Map<String, ?> body;
        public final Map<String, String> query;
        public final String captchaToken;

        public FunctionCallOptions() {
            this("POST", Collections.emptyMap(), null, null);
        }

        public FunctionCallOptions(String method, Map<String, ?> body, Map<String, String> query) {
            this(method, body, query, null);
        }

        public FunctionCallOptions(String method, Map<String, ?> body, Map<String, String> query,
                String captchaToken) {
            this.method = method != null ? method : "POST";
            this.body = body != null ? body : Collections.emptyMap();
            this.query = query;
            this.captchaToken = captchaToken;
        }
    }

    private final HttpClient httpClient;

    public FunctionsClient(HttpClient httpClient) {
        this.httpClient = httpClient;
    }

    public Object call(String path) throws EdgeBaseError {
        return call(path, new FunctionCallOptions());
    }

    public Object call(String path, FunctionCallOptions options) throws EdgeBaseError {
        String normalizedPath = "/functions/" + path;
        String method = options != null ? options.method.toUpperCase() : "POST";
        Map<String, ?> body = options != null && options.body != null ? options.body : Collections.emptyMap();
        Map<String, String> query = options != null ? options.query : null;
        String captchaToken = options != null ? options.captchaToken : null;
        if (captchaToken != null && (captchaToken.isEmpty() || captchaToken.length() > 2048)) {
            throw new IllegalArgumentException("captchaToken must be non-empty and at most 2048 characters");
        }

        switch (method) {
            case "GET":
                return httpClient.get(normalizedPath, query, captchaToken);
            case "PUT":
                return httpClient.put(normalizedPath, body, captchaToken);
            case "PATCH":
                return httpClient.patch(normalizedPath, body, captchaToken);
            case "DELETE":
                return captchaToken == null
                        ? httpClient.delete(normalizedPath)
                        : httpClient.deleteWithCaptchaToken(normalizedPath, captchaToken);
            case "POST":
            default:
                return httpClient.post(normalizedPath, body, captchaToken);
        }
    }

    public Object get(String path) throws EdgeBaseError {
        return get(path, null);
    }

    public Object get(String path, Map<String, String> query) throws EdgeBaseError {
        return call(path, new FunctionCallOptions("GET", null, query));
    }

    public Object post(String path) throws EdgeBaseError {
        return post(path, Collections.emptyMap());
    }

    public Object post(String path, Map<String, ?> body) throws EdgeBaseError {
        return call(path, new FunctionCallOptions("POST", body, null));
    }

    public Object put(String path, Map<String, ?> body) throws EdgeBaseError {
        return call(path, new FunctionCallOptions("PUT", body, null));
    }

    public Object patch(String path, Map<String, ?> body) throws EdgeBaseError {
        return call(path, new FunctionCallOptions("PATCH", body, null));
    }

    public Object delete(String path) throws EdgeBaseError {
        return call(path, new FunctionCallOptions("DELETE", null, null));
    }
}

File: packages/sdk/kotlin/client/src/commonMain/kotlin/dev/edgebase/sdk/client/FunctionsClient.kt
package dev.edgebase.sdk.client

import dev.edgebase.sdk.core.HttpClient

data class FunctionCallOptions(
    val method: String = "POST",
    val body: Map<String, Any?>? = null,
    val query: Map<String, String>? = null,
    val captchaToken: String? = null
)

class FunctionsClient(
    private val httpClient: HttpClient
) {
    suspend fun call(path: String, options: FunctionCallOptions = FunctionCallOptions()): Any? {
        val normalizedPath = "/functions/$path"
        options.captchaToken?.let {
            require(it.isNotEmpty() && it.length <= 2_048) {
                "captchaToken must be non-empty and at most 2048 characters"
            }
        }
        return when (options.method.uppercase()) {
            "GET" -> httpClient.get(normalizedPath, options.query, options.captchaToken)
            "PUT" -> httpClient.put(normalizedPath, options.body ?: emptyMap(), options.captchaToken)
            "PATCH" -> httpClient.patch(normalizedPath, options.body ?: emptyMap(), options.captchaToken)
            "DELETE" -> httpClient.delete(normalizedPath, captchaToken = options.captchaToken)
            else -> httpClient.post(normalizedPath, options.body ?: emptyMap(), options.captchaToken)
        }
    }

    suspend fun get(path: String, query: Map<String, String>? = null): Any? =
        call(path, FunctionCallOptions(method = "GET", query = query))

    suspend fun post(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        call(path, FunctionCallOptions(method = "POST", body = body))

    suspend fun put(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        call(path, FunctionCallOptions(method = "PUT", body = body))

    suspend fun patch(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        call(path, FunctionCallOptions(method = "PATCH", body = body))

    suspend fun delete(path: String): Any? =
        call(path, FunctionCallOptions(method = "DELETE"))
}
